tests/pulse/test_parametrized_evolution.py

```diff
@@ -36,18 +36,82 @@ class MyOp(qml.RX):  # pylint: disable=too-few-public-methods
     has_diagonalizing_gates = False
 
 
+def amp0(p, t):
+    return p * t
+
+
+def amp1(p, t):
+    return p[0] * t + p[1]
+
+
+H0 = qml.PauliX(1) + amp0 * qml.PauliZ(0) + amp0 * qml.PauliY(1)
+params0_ = [0.5, 0.5]
+
+H1 = qml.PauliX(1) + amp0 * qml.PauliZ(0) + amp1 * qml.PauliY(1)
+params1_ = (0.5, [0.5, 0.5])
+
+example_pytree_evolutions = [
+    qml.pulse.ParametrizedEvolution(H0),
+    qml.pulse.ParametrizedEvolution(H0, params0_),
+    qml.pulse.ParametrizedEvolution(H0, t=0.3),
+    qml.pulse.ParametrizedEvolution(H0, params0_, t=0.5),
+    qml.pulse.ParametrizedEvolution(H0, params0_, t=[0.5, 1.0]),
+    qml.pulse.ParametrizedEvolution(H0, params0_, t=0.5, return_intermediate=True),
+    qml.pulse.ParametrizedEvolution(
+        H0, params0_, t=0.5, return_intermediate=True, complementary=True
+    ),
+    qml.pulse.ParametrizedEvolution(
+        H0, params0_, t=0.5, return_intermediate=True, complementary=True, atol=1e-4, rtol=1e-4
+    ),
+    qml.pulse.ParametrizedEvolution(
+        H0,
+        params0_,
+        t=0.5,
+        return_intermediate=True,
+        complementary=True,
+        atol=1e-4,
+        rtol=1e-4,
+        dense=True,
+    ),
+    qml.pulse.ParametrizedEvolution(H1, params1_, t=0.5),
+    qml.pulse.ParametrizedEvolution(H1, params1_, t=0.5, return_intermediate=True),
+]
+
+@pytest.mark.parametrize("evol", example_pytree_evolutions)
+def test_trivial_equality(evol):
+    """Test equality method for the trivial case of having the same operators"""
+    assert evol == evol
+
+
+def test_equality_different_datatypes():
+    """Test equality method for the case of same data but different datatypes"""
+    p0 = [0.5, 0.5]
+    p1 = (0.5, 0.5)
+    evol1 = qml.pulse.ParametrizedEvolution(H0, p0, t=0.5)
+    evol2 = qml.pulse.ParametrizedEvolution(H0, p1, t=0.5)
+    assert evol1 == evol2
+
+@pytest.mark.jax
+class TestPytree:
+    """Testing pytree related functionality"""
+
+    @pytest.mark.parametrize("evol", example_pytree_evolutions)
+    def test_flatten_unflatten_identity(self, evol):
+        """Test that flattening and unflattening is yielding the same parametrized evolution"""
+        assert evol._unflatten(*evol._flatten()) == evol
+
+
 @pytest.mark.jax
-@pytest.mark.xfail  # TODO: fix by story 50356
```

Comment:
When I run the test in a python shell it works, but in pytest I get the following error. What is this pickling business about?

```pycon
pennylane/ops/functions/assert_valid.py:284: in assert_valid
    _check_pickle(op)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

op = ParametrizedEvolution(0.5, wires=[0])

    def _check_pickle(op):
        """Check that an operation can be dumped and reloaded with pickle."""
>       pickled = pickle.dumps(op)
E       AttributeError: Can't pickle local object 'test_standard_validity.<locals>.f1'

pennylane/ops/functions/assert_valid.py:204: AttributeError
```

Reply:
So `f1` is a local function, so its not pickleable.  

Since we know that pickling in finicky and a lot of things can cause it to fail, you can just do `assert_valid(ev, skip_pickle=True)`.  

Or you can move the function out of the function to be global.

Reply:
Thanks!